src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs

```diff
@@ -334,50 +334,125 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
         /// </summary>
         public static IEnumerable<ResolverPackage> FindCircularDependency(IEnumerable<ResolverPackage> solution)
         {
+            // to keep track of visited packages to avoid processing them again
+            var visitedPackages = new Dictionary<string, HashSet<ResolverPackage>>();
+
+            var finalResult = Enumerable.Empty<ResolverPackage>();
+
             // check each package to see if it is part of a loop, sort by id to keep the result deterministic
-            foreach (var package in solution.OrderBy(package => package.Id, StringComparer.OrdinalIgnoreCase))
+            foreach (var package in solution)
             {
-                var result = FindCircularDependency(package, Enumerable.Empty<string>(), solution);
-
-                if (result.Any())
+                var result = FindCircularDependency(package, solution, visitedPackages);
+                if (result.Count() > finalResult.Count())
                 {
-                    // loop found
-                    return result;
+                    // loop found which also has more packages then existing loop
+                    finalResult = result;
                 }
             }
 
-            // no loops detected
-            return Enumerable.Empty<ResolverPackage>();
+            return finalResult;
         }
 
-        private static IEnumerable<ResolverPackage> FindCircularDependency(ResolverPackage package, IEnumerable<string> parents, IEnumerable<ResolverPackage> solution)
+        private static IEnumerable<ResolverPackage> FindCircularDependency(ResolverPackage package, IEnumerable<ResolverPackage> solution, Dictionary<string, HashSet<ResolverPackage>> visitedPackages)
```

Comment:
Instead of taking IEnumerable<ResolverPackage> and looping each time to find the dependency, index this once outside of this method into a: SortedDictionary<string, SortedSet<string>> to get package id -> dependency ids   this will also take care of the sorting for you so that this is deterministic.